### PR TITLE
Check Performance exists before using it

### PR DESCRIPTION
**Pull request by benjamingr**
#### Purpose (TL;DR) - mandatory

Fixes https://github.com/sinonjs/lolex/issues/171#issuecomment-389487264 since apparently some environments expose `global.performance` but not `global.Performance`. 

#### Solution  - optional

Guard checking performance with `global.Performance`.

### CODE REVIEW

**Comment by SimenB** [comment]
This means environments with `performance.now` but no `Performance` will _not_ be mocked, right?

Would it be better to guard around specifically https://github.com/sinonjs/lolex/blob/f7d11449c955556652d662b463f9e42903d41903/src/lolex-src.js#L751-L759 but leave `performance.now` to be mocked?

**Comment by benjamingr** [comment]
@SimenB Looking at https://developer.mozilla.org/en-US/docs/Web/API/Performance and https://developer.mozilla.org/en-US/docs/Web/API/Performance/now I don't really understand what environments do one but not the other.

**Comment by benjamingr** [comment]
BTW - If you feel strongly about it I'll gladly edit the PR. I just want a better understanding of where this problem bites people.

Another option is to get `Performance` by `performance.__proto__` (performance doesn't exist in IE which doesn't have __proto__ - we can also do Object.getPrototypeOf).

**Comment by SimenB** [comment]
Doesn't impact me, just throwing out suggestions 🙂

**Comment by benjamingr** [comment]
Ok, let's wait to see if the user who complained shows up to provide an environment we can produce a failing case for then :)

**Comment by Kyr** [comment]
Please, not only check is Performance or performance present, this [fragment]( https://github.com/sinonjs/lolex/blame/f7d11449c955556652d662b463f9e42903d41903/src/lolex-src.js#L751-L759) cause serious problem on node, 8, 10, whatever due to Performance does not present, and more over:
```
> perf_hooks.performance.__proto__
Performance {}
> perf_hooks.performance.prototype
undefined
```
so any sinon's fake timers does not work

**Comment by benjamingr** [comment]
@Kyr thanks for stepping in and contributing. Note that `Performance` is the performance prototype - which is the same as `performance.__proto__`

So generally since `performance instanceof Performance` that means `performance.__proto__ === Performance.prototype`. Note the uppercase `P`

**Comment by benjamingr** [comment]
Also, in Node 8 performance isn't exported a global which would mean the entire piece of code wouldn't get hit at all initially. If you have code reproducing the issue I'd be very interested in seeing it.

Thanks again.

**Comment by Kyr** [comment]
Thanks for explanation @benjamingr , but may I apoint you to something another
image

node v8.11.2

**Comment by benjamingr** [comment]
@Kyr thanks a lot. Can you tell us what environment this is running in? (Node?)

Can you please share:

 - What version is running (Node 8.x?)
 - What's the value of `_global`? 
 - What's the value of `_global.performance`?
 - Is this a JSDOM test (or something similar to a JSDOM test?)

**Comment by Kyr** [comment]
@benjamingr I apologize for being misled, I completely forgot about the years made that custom jsdom setup which broke everything.

After removing already unnecessary patches, everything started to work fine.

**Comment by benjamingr** [comment]
> After removing already unnecessary patches, everything started to work fine.

I would still like to make sure that we don't break user deployments. I'm a firm believer in lolex not breaking peoples' code and I'll do better in the future.

Thanks for the report - that's helpful. 

@SimenB given the above - what do you think we should do?

**Comment by benjamingr** [comment]
Looks like this is an "overriding globals" issue https://github.com/jsdom/jsdom/issues/2231 - given that and not a misconfiguration I'm going to go ahead and go with @SimenB's suggested approach

**Comment by benjamingr** [comment]
Updated, PTAL

File: src/lolex-src.js [hunk 1]
```diff
@@ -750,13 +751,15 @@ function withGlobal(_global) {
 
             var proto = _global.Performance.prototype;
```
Comment (SimenB) [inline src/lolex-src.js]: Missing a guard here

Comment (SimenB) [reply src/lolex-src.js]: Can probably reuse `proto` on line 756

Comment (benjamingr) [reply src/lolex-src.js]: Yeah, makes sense

**Pull request merged**